tests/unit/s2n_pkey_test.c

```diff
@@ -13,158 +13,175 @@
-            EXPECT_FAILURE_WITH_ERRNO(s2n_cert_chain_and_key_load_pem_bytes(chain_and_key, rsa_pss_cert_chain_pem, rsa_pss_cert_chain_len, ecdsa_private_key_pem, ecdsa_private_key_len),
-                    S2N_ERR_KEY_MISMATCH);
-            EXPECT_SUCCESS(s2n_cert_chain_and_key_free(chain_and_key));
-
-            EXPECT_NOT_NULL(chain_and_key = s2n_cert_chain_and_key_new());
-            EXPECT_FAILURE_WITH_ERRNO(s2n_cert_chain_and_key_load_pem_bytes(chain_and_key, ecdsa_cert_chain_pem, ecdsa_cert_chain_len, rsa_pss_private_key_pem, rsa_pss_private_key_len),
-                    S2N_ERR_KEY_MISMATCH);
-            EXPECT_SUCCESS(s2n_cert_chain_and_key_free(chain_and_key));
-        }
+                    s2n_pkey_encrypt(rsa_pss_pkeys.priv_key, &in, &out),
+                    S2N_ERR_UNIMPLEMENTED);
+            EXPECT_FAILURE_WITH_ERRNO(
+                    s2n_pkey_decrypt(&rsa_pss_pkeys.pub_key, &in, &out),
+                    S2N_ERR_UNIMPLEMENTED);
+        };
+
+        /* Test: supported for RSA */
+        {
+            uint8_t message_bytes[] = "hello world";
+            struct s2n_blob message = { 0 };
+            EXPECT_SUCCESS(s2n_blob_init(&message, message_bytes, sizeof(message_bytes)));
+
+            const char ciphertext_hex[] =
+                    "38 ab 9c 83 57 17 13 46 3d 5b 6f c6 44 30 8e 40 78 ac d2 "
+                    "c5 69 98 8b 99 51 14 23 c6 af 98 67 6f 58 d5 28 9e 23 12 "
+                    "6f 5b 73 56 d9 f9 b2 10 4f d6 53 5c 70 a5 a6 c6 a2 53 83 "
+                    "7d ec 3b 7c c0 ff 6a 30 8d 98 a3 6c be c6 e3 b4 4d b6 3d "
+                    "cc 94 67 f8 24 9a 91 13 52 57 02 0c d9 5a 00 98 1f ff df "
+                    "5e 47 e0 ca 15 87 f0 92 1e 95 af ef 49 b2 6b f2 b6 be d5 "
+                    "3b 65 d3 94 92 f5 c1 f0 65 56 20 85 7f 18 95 a5 d9 e7 6c "
+                    "43 07 dd 5d 03 60 ac 4d c5 a0 c8 3d f9 99 24 fc 30 8f c2 "
+                    "66 9d df 5c 80 90 a7 c5 7a 37 ee be 1d 30 a7 a3 67 73 ae "
+                    "7d ee 64 37 22 77 9a a5 0d 47 f0 a5 50 ee 85 82 2e 88 32 "
+                    "e9 0b bc 25 5f 09 b7 d3 13 58 88 84 9d 07 03 5e 37 6b af "
+                    "08 56 14 fd 64 58 29 5b 81 a5 72 72 62 5d c1 72 bb 13 76 "
+                    "b6 17 96 7b d9 87 ec 49 71 dc 33 3e b2 f5 76 54 ad 13 ed "
+                    "23 1c 34 53 d1 12 03 be f6";
+            S2N_BLOB_FROM_HEX(ciphertext, ciphertext_hex);
+
+            /* Test: legacy RSA decryption accepts known good value */
+            {
+                DEFER_CLEANUP(struct s2n_blob output = { 0 }, s2n_free);
+                EXPECT_SUCCESS(s2n_alloc(&output, message.size));
+
+                EXPECT_SUCCESS(s2n_rsa_decrypt(rsa_pkeys.priv_key, &ciphertext, &output));
+                EXPECT_BYTEARRAY_EQUAL(output.data, message.data, message.size);
+            };
+
+            /* Test: decryption works for known good value */
```

Comment:
Can you also add a bad value test for decryption? Seems like that is missing.

Reply:
Oh, good call. A unit test should definitely yell at you if you try to make failures actually fail instead of the unintuitive "just return random" we do for Bleichenbacher.